Ditch pathlib and fix Windows bugs

kgadams opened the pull request:
Description
-------------

This PR gets rid of the pathlib, replacing all calls with calls to the std library.

No change in generated files, no new warnings, no test errors. A couple of new Unittests.

- Fixes #1001
- Fixes #1037
- Fixes #1060

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Version of Go used when building/testing:
---------------------------------------------

- [ ] 1.22
- [ ] 1.23
- [x] 1.24

How Has This Been Tested?
---------------------------

I tested the code on Mac running the defaults from the Taskfile, and I tested the reproducers from the three issues below on Windows. All of them work now.

Testing on Windows does not work from the taskfile, it assumes a lot of unix tools. I ran the tests manually as far as possible.

Checklist
-----------

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Commit:
Ditch pathlib, it does not support Windows

Commit:
Windows porting

Commit:
Fix e2etest on windows

LandonTClipp commented on template_funcs/funcmap.go:
```diff
@@ -58,9 +58,9 @@ var FuncMap = template.FuncMap{
-	"base":     filepath.Base,
-	"clean":    filepath.Clean,
-	"dir":      filepath.Dir,
+	"base":     func(path string) string { return filepath.ToSlash(filepath.Base(path)) },
+	"clean":    func(path string) string { return filepath.ToSlash(filepath.Clean(path)) },
+	"dir":      func(path string) string { return filepath.ToSlash(filepath.Dir(path)) },
```
Is this correct? Why would you want this? If I'm reading correctly this will replace all `\` characters in the Windows path with `/`.

kgadams replied:
First of all, it makes the unittests pass.
Joke aside: in all of my coding I keep pathes with slashes on all platforms. It works just fine on Windows.
Also it makes life easier for the writers of the templates: they can rely on all pathes having slashes as separator, so you don't have to write template variants for different platforms.

LandonTClipp replied:
Interesting, so the tests don't pass unless you use `filepath.ToSlash`? I would be curious to know why. I am just suspicious of this change because someone might, for example, want to read the contents of a file (copyright text, for example), the path of which might need to pass through one of these functions. I don't know if Windows would understand you trying to open a file with `/` characters.

Kind of sad that the Go Playground does not seem to have Windows environments I can test this out with.

kgadams replied:
The unittests in `funcmap_test.go` for these template functions expect the result to contain slashes (`/home/file.txt`). Without the change, you get `\home\file.txt` as result, from the `clean` template function test. So you would have to have OS-specific unittests.
Believe you me, nobody wants this.
And go APIs on Windows work just fine with slashes (C APIs too for that matter). Only when you build actual commandlines for consumption by `cmd.exe` do you need backslashes.

kgadams replied:
BTW: everywhere pathes are now stored in the coding in my PR, they are stored after conversion to slashes. As mentioned, opening files etc. works just fine with these on Windows.

LandonTClipp replied:
I see, thank you for explaining. I am obviously wholly unfamiliar with Windows development! I'll approve and merge/release this in a few. Thanks again for the effort, I needed a Windows dev to help with these problems.

LandonTClipp reviewed (commented):
Looks great overall, thank you! Just one question on the template func changes.

I wonder if you can try adding windows to the testing matrix. I am 90% confident it won't just work out of the box but I am curious to try. If it's too much trouble for you, we can try it in a different PR.

kgadams commented:
> Looks great overall, thank you! Just one question on the template func changes.
> 
> I wonder if you can try adding windows to the testing matrix. I am 90% confident it won't just work out of the box but I am curious to try. If it's too much trouble for you, we can try it in a different PR.

I am not familiar with the testing infrastructure on the public github, normally I work in a corporate github, where this works very much in a different way.
That being said: your scripting already has problems on Windows in relying on a unix shell and on unix commandline utilities... Just running `task` fails miserably (`xargs not found` etc.).

Commit:
Port taskfile to windows

Commit:
Fix taskfile

Commit:
Keep lf even on windows

kgadams commented:
> > Looks great overall, thank you! Just one question on the template func changes.
> > I wonder if you can try adding windows to the testing matrix. I am 90% confident it won't just work out of the box but I am curious to try. If it's too much trouble for you, we can try it in a different PR.
> 
> I am not familiar with the testing infrastructure on the public github, normally I work in a corporate github, where this works very much in a different way. That being said: your scripting already has problems on Windows in relying on a unix shell and on unix commandline utilities... Just running `task` fails miserably (`xargs not found` etc.).

Ok, I added `windows-latest` into the test matrix and fixed the remaining issues. Two forgotten `file.Close()` and some updates in the `Taskfile.yml` to make it work on Windows. It turns out that bash is available on the Windows runners, so the e2e scripts work.

LandonTClipp commented:
Wow you've gone above and beyond, that's spectacular. Thank you!
